server/user_db.go

```diff
@@ -0,0 +1,238 @@
+	db.keys[name] = index
+
+	return u, nil
+}
+
+func (db *userDb) Get(s string) (*pb.User, error) {
+	if v, ok := db.keys[s]; ok {
+		entry := db.users[int(v)]
+		if !entry.deleted {
+			return entry.user, nil
+		}
+	}
+
+	return nil, status.Errorf(codes.NotFound, "A user with name %s not found.", s)
+}
+
+func (db *userDb) Update(u *pb.User, f *field_mask.FieldMask) (*pb.User, error) {
+	if f != nil && len(f.GetPaths()) > 0 {
+		return nil, status.Error(
+			codes.Unimplemented,
+			"Field masks are currently not supported.")
+	}
+
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	err := db.validate(u)
+	if err != nil {
+		return nil, err
+	}
+
+	if i, ok := db.keys[u.GetName()]; ok {
+		entry := db.users[int(i)]
+		if !entry.deleted {
+			// Validate Unique Fields.
+			hasSameName := func(x *pb.User) bool {
+				return x != entry.user && u.GetDisplayName() == x.GetDisplayName()
+			}
+			if db.anyUser(hasSameName) {
+				return nil, status.Errorf(
+					codes.AlreadyExists,
+					"A user with display_name %s already exists.",
+					u.GetDisplayName())
+			}
+
+			hasSameEmail := func(x *pb.User) bool {
+				return x != entry.user && u.GetEmail() == x.GetEmail()
+			}
+			if db.anyUser(hasSameEmail) {
+				return nil, status.Errorf(
+					codes.AlreadyExists,
+					"A user with the email %s already exists.",
+					u.GetEmail())
+			}
+
+			// Update store.
+			db.users[int(i)] = userEntry{user: u, deleted: false}
+			return u, nil
+		}
+	}
+
+	return nil, status.Errorf(
+		codes.NotFound,
+		"A user with name %s not found.", u.GetName())
+}
+
+func (db *userDb) Delete(s string) error {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	i, ok := db.keys[s]
+
+	if !ok {
+		return status.Errorf(
+			codes.NotFound,
+			"A user with name %s not found.", s)
+	}
+
+	entry := db.users[int(i)]
+	db.users[int(i)] = userEntry{user: entry.user, deleted: true}
+
+	return nil
+}
+
+func (db *userDb) List(pageSize int32, pageToken string) (*pb.ListUsersResponse, error) {
+	start, err := db.token.GetIndex(pageToken)
+	if err != nil {
+		return nil, err
+	}
+
+	offset := 0
+	users := []*pb.User{}
+	for _, entry := range db.users[start:] {
+		offset++
+		if entry.deleted {
+			continue
+		}
+		users = append(users, entry.user)
+		if len(users) >= int(pageSize) {
+			break
+		}
+	}
+
+	nextToken := ""
+	if start+offset < len(db.users) {
+		nextToken = db.token.ForIndex(start + offset)
+	}
+
+	return &pb.ListUsersResponse{Users: users, NextPageToken: nextToken}, nil
+}
+
+func (db *userDb) validate(u *pb.User) error {
+	// Validate Required Fields.
+	if u.GetDisplayName() == "" {
+		return status.Errorf(
+			codes.InvalidArgument,
+			"The field `display_name` is required.")
+	}
+	if u.GetEmail() == "" {
+		return status.Errorf(
+			codes.InvalidArgument,
+			"The field `email` is required.")
+	}
+	return nil
+}
+
+func (db *userDb) anyUser(f func(*pb.User) bool) bool {
+	for _, entry := range db.users {
+		if entry.deleted {
+			continue
```

Comment:
Considering there is no logic following the `if f()` check, might be cleaner to do 
```
if !entry.deleted && f(entry.user) {
  return true
}
```
Not blocking, just a suggestion!

Reply:
lol woops! Fixed!